win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1167,214 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
+            RtlCopyMemory(szzFontLink, s_szzDefFontLink, SZZ_GetSize(s_szzDefFontLink));
+        else
+            RtlCopyMemory(szzFontLink, s_szzDefFixedPitchFontLink,
+                          SZZ_GetSize(s_szzDefFixedPitchFontLink));
+    }
+
+    // Ensure double-NUL-terminated
+    szzFontLink[_countof(szzFontLink) - 1] = UNICODE_NULL;
+    szzFontLink[_countof(szzFontLink) - 2] = UNICODE_NULL;
+
+    FontLinkSize = SZZ_GetSize(szzFontLink);
+    pszzFontLink = ExAllocatePoolWithTag(PagedPool, FontLinkSize, TAG_FONT);
+    if (!pszzFontLink)
+        return STATUS_NO_MEMORY;
+    RtlCopyMemory(pszzFontLink, szzFontLink, FontLinkSize);
+    pChain->pszzFontLink = pszzFontLink;
+
+    return STATUS_SUCCESS;
+}
+
+static PFONTLINK
+FontLink_CreateOrGet(
+    _In_ const LOGFONTW *plfBase,
+    _In_ LPCWSTR pszLink)
+{
+    LPWSTR pch1, pch2;
+    LOGFONTW lf;
+    PFONTLINK pLink;
+
+    lf = *plfBase;
+    lf.lfCharSet = DEFAULT_CHARSET;
+
+    // pszLink: "<FontFileName>,<FaceName>[,...]"
+    pch1 = wcschr(pszLink, L',');
+    if (!pch1)
+    {
+        DPRINT1("%S\n", pszLink);
+        return NULL;
+    }
+    ++pch1;
+
+    pch2 = wcschr(pch1, L',');
+    if (pch2)
+        RtlStringCchCopyNW(lf.lfFaceName, _countof(lf.lfFaceName), pch1, pch2 - pch1);
+    else
+        RtlStringCchCopyW(lf.lfFaceName, _countof(lf.lfFaceName), pch1);
+
+    SubstituteFontRecurse(&lf);
+    DPRINT1("lfFaceName: %S\n", lf.lfFaceName);
+
+    pLink = ExAllocatePoolZero(PagedPool, sizeof(FONTLINK), TAG_FONT);
+    if (!pLink)
+    {
+        DPRINT1("!pLink\n");
+        return NULL;
+    }
+
+    pLink->LogFont = lf;
+
+    if (_wcsicmp(plfBase->lfFaceName, lf.lfFaceName) == 0)
+        pLink->bIgnoreThis = TRUE;
+
+    return pLink;
+}
+
+static NTSTATUS
+FontLink_Chain_Populate(
+    _Inout_ PFONTLINK_CHAIN pChain)
+{
+    NTSTATUS Status;
+    PFONTLINK pLink;
+    LOGFONTW lfBase;
+    PTEXTOBJ pTextObj = pChain->pBaseTextObj;
+    PFONTGDI pFontGDI;
+    PZZWSTR pszzFontLink;
+    WCHAR *pchLink;
+    PFONTLINK_CACHE pLinkCache;
+
+    InitializeListHead(&pChain->FontLinkList);
+
+    if (!pTextObj->TextFace[0])
+        return STATUS_INVALID_PARAMETER;
+
+    lfBase = pTextObj->logfont.elfEnumLogfontEx.elfLogFont;
+    pFontGDI = ObjToGDI(pTextObj->Font, FONT);
+    lfBase.lfHeight = pFontGDI->lfHeight;
+    lfBase.lfWidth = pFontGDI->lfWidth;
+
+    // lfBase.lfFaceName <-- pTextObj->TextFace (if lfFaceName was empty)
+    if (!lfBase.lfFaceName[0])
+        RtlStringCchCopyW(lfBase.lfFaceName, _countof(lfBase.lfFaceName), pTextObj->TextFace);
+
+    pLinkCache = FontLink_FindCache(&lfBase);
+    if (pLinkCache)
+    {
+        *pChain = pLinkCache->Chain;
+        return STATUS_SUCCESS;
+    }
+
+    // Load FontLink entry from registry
+    Status = FontLink_Chain_LoadReg(pChain, &lfBase);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    pszzFontLink = pChain->pszzFontLink;
+    pchLink = pszzFontLink;
```

Comment:
```suggestion
    pchLink = pChain->pszzFontLink;
```
and the intermediate `pszzFontLink` could be removed.

Reply:
Removed.